planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+
+* Maybe introduce a new concept of `ILinqDialect` where the `Expression` parsing would just detect *what* logical thing it finds (like `!BoolProperty`), and turns around and calls this `ILinqDialect` to get at a `WhereFragment` or whatever. This way we could ready ourselves to support an alternative json/sql dialect around JSONPath for Postgresql v12+ and later for Sql Server vNext. I think this would fit into the theme of making the Linq support more modular. It *should* make the Linq support easier to unit test as we go. Before we do anything with this, let's take a deep look into the EF Core internals and see how they handle this issue
+
+* Consider replacing the `SelectMany()` implementation with *Common Table Expression* sql statements. That might do a lot to simplify the internal mechanics. Could definitely get us to an n-deep model.
+
+* Do the [Json streaming story](https://github.com/JasperFx/marten/issues/585) because it should be compelling, especially as part of the readside of a CQRS architecture using Marten's event store functionality. 
+
+## Dynamic Code Generation
+
+* Adopt LamarCodeGeneration for building out the dynamic code. The current Expression building is wonky, and it's getting creaky. This will make the document metadata functionality
+  easier to build. We can do this either by continuing to build Expressions and compiling to Funcs, or we could re-introduce code compilation via LamarCompiler (which was originally
+  built into Marten pre-V1)
+
+**The potential downside is that this doesn't work well at all with `internal` properties, so we'll have to watch that**
+
+
+## Partial Updates
+
+Use native postgresql partial JSON updates wherever possible. Let's do a perf test on that first though.
+
+## Event Sourcing
+
+There's an existing conversation on GitHub about the [event sourcing improvements for v4](https://github.com/JasperFx/marten/issues/1307).
+
+### Event Store Metadata
+
+See the earlier conversation about document metadata up above, and also the next section on schema generation for the event store.
+
+What if we allow for custom event wrapper (what's now `IEvent` and `Event<T>`) types, and event stream types? So you could use custom
+metadata information right on those types? And possibly even put some logic into event stream creation to capture stream
+metadata?
+
+If so, those types could be something like this:
+
+```
+    public enum StreamState
+    {
+        CreateNew,
+        Append,
+        AppendOrCreate,
+        Pending,
+        History
+    }
+    
+    // This really represents a segment of the underlying stream
+    public abstract class EventStreamBase<TKey, TEvent, TEventBase> where TEvent : IEvent<TEventBase>
```

Comment:
How this class would be used?

Reply:
Internally mostly, but this would represent part of a stream in projection operations. I'm also thinking this would be used to map metadata concerns per stream

Reply:
Definitely, I'd need some sample to say yay or nay ;)